### ISSUE TITLE
Rounding errors in masked jnp.sum

### ISSUE THREAD

**Issue by RobRomijnders** [google/jax#6553]
Dear JAX team,

Thanks for open sourcing this library. 

Apologies for my lack of knowledge in numerical computing. My question concerns how multiplies by zero are handled. I am encountering inconsistent rounding when summing large matrices, e.g. `jnp.sum(matrix[:, :slice])` has a different result than `jnp.sum(matrix*mask)`.

Code to reproduce this behavior:

```python
mask = jnp.concatenate((np.ones((1000, 1000)), np.zeros((1000, 1000))), axis=1)
num_match = 0
for _ in range(100):
  matrix = np.random.random((1000, 2000))

  sum_sliced = jnp.sum(matrix[:, :1000])
  sum_masked = jnp.sum(matrix * mask)

  num_match += int(sum_sliced == sum_masked)

  print(f"{sum_sliced:15.10f}")
  print(f"{sum_masked:15.10f}")
  print()

print(f"Matches in {num_match}/100 cases")
```

This snippet will print:

```
  499409.312500
  499409.312500

  499967.500000
  499967.500000

  500077.562500
  500077.531250

...

Matches in 51/100 cases
```

Apparently, there's a rounding error happening in about half of the cases. Is this behavior expected? 

Kind regards,
Rob

**Comment by jakevdp** [comment]
In general, you should not expect exact equality for different expressions of the same operation in floating point. As a simple example, take this sum in numpy:
```python
import numpy as np
x = np.random.rand(2000)
print(x.sum() == x[::-1].sum())
# False
```
The behavior you see above I think comes from a similar cause: the XLA summation adds the numbers in a different order in the two expressions, but nevertheless you should see that the results agree to within normal floating point rounding errors.

**Comment by jakevdp** [comment]
Closing, as I think this is expected behavior.
